zephyr/lib/alloc.c

```diff
@@ -145,14 +145,89 @@ static bool is_l3_heap_pointer(void *ptr)
 	uintptr_t l3_heap_start = get_l3_heap_start();
 	uintptr_t l3_heap_end = l3_heap_start + get_l3_heap_size();
 
-	if (is_cached(ptr))
-		ptr = z_soc_uncached_ptr((__sparse_force void __sparse_cache *)ptr);
-
 	if ((POINTER_TO_UINT(ptr) >= l3_heap_start) && (POINTER_TO_UINT(ptr) < l3_heap_end))
 		return true;
 
 	return false;
 }
+
+/**
+ * Flush cached L3 heap management data.
+ */
+static inline void l3_heap_flush(struct k_heap *h)
+{
+	if (h == &l3_heap)
+		sys_cache_data_flush_and_invd_range((void *)get_l3_heap_start(),
+						    get_l3_heap_size());
+}
+
+/**
+ * Invalidate cached L3 heap management data.
+ */
+static inline void l3_heap_invalidate(struct k_heap *h)
+{
+	if (h == &l3_heap)
+		sys_cache_data_invd_range((void *)get_l3_heap_start(),
+					  get_l3_heap_size());
+}
+
+static void *l3_heap_alloc_aligned(struct k_heap *h, size_t min_align, size_t bytes)
+{
+	k_spinlock_key_t key;
+	void *ret;
+#if CONFIG_SYS_HEAP_RUNTIME_STATS && CONFIG_IPC_MAJOR_4
+	struct sys_memory_stats stats;
+#endif
+
+	key = k_spin_lock(&h->lock);
+	l3_heap_invalidate(h);
+	ret = sys_heap_aligned_alloc(&h->heap, min_align, bytes);
+	l3_heap_flush(h);
+	k_spin_unlock(&h->lock, key);
+
+#if CONFIG_SYS_HEAP_RUNTIME_STATS && CONFIG_IPC_MAJOR_4
+	sys_heap_runtime_stats_get(&h->heap, &stats);
+	tr_info(&zephyr_tr, "heap allocated: %u free: %u max allocated: %u",
+		stats.allocated_bytes, stats.free_bytes, stats.max_allocated_bytes);
+#endif
+
+	return ret;
+}
+
+static void __sparse_cache *l3_heap_alloc_aligned_cached(struct k_heap *h,
+							 size_t min_align, size_t bytes)
+{
+	void __sparse_cache *ptr;
+
+	/*
+	 * Zephyr sys_heap stores metadata at start of each
+	 * heap allocation. To ensure no allocated cached buffer
+	 * overlaps the same cacheline with the metadata chunk,
+	 * align both allocation start and size of allocation
+	 * to cacheline. As cached and non-cached allocations are
+	 * mixed, same rules need to be followed for both type of
+	 * allocations.
+	 */
+	min_align = MAX(PLATFORM_DCACHE_ALIGN, min_align);
+	bytes = ALIGN_UP(bytes, min_align);
+
+	ptr = (__sparse_force void __sparse_cache *)l3_heap_alloc_aligned(h, min_align, bytes);
+
+	if (ptr)
+		ptr = z_soc_cached_ptr((__sparse_force void *)ptr);
+
+	return ptr;
+}
+
+static void l3_heap_free(struct k_heap *h, void *mem)
+{
+	k_spinlock_key_t key = k_spin_lock(&h->lock);
+	sys_cache_data_flush_and_invd_range(mem, get_l3_heap_size());
```

Comment:
this looks wrong - please either use `sys_cache_data_*_()` directly or your above wrappers consistently everywhere. Then - you manipulate cache for a range, starting at *some* address in L3 memory, but with size equal to entire L3 memory? So you might try to manipulate cache for almost an entire L3 size beyond the L3 heap range? Then - I don't think this is needed at all. Why write back user data if the memory is freed? The data is discarded. Why care to invalidate it here, maybe better do it when allocating a cached buffer

Reply:
fixed